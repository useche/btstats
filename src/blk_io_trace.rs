use crate::bindings::{
    __BLK_TA_ABORT, __BLK_TA_BACKMERGE, __BLK_TA_BOUNCE, __BLK_TA_COMPLETE, __BLK_TA_DRV_DATA,
    __BLK_TA_FRONTMERGE, __BLK_TA_GETRQ, __BLK_TA_INSERT, __BLK_TA_ISSUE, __BLK_TA_PLUG,
    __BLK_TA_REMAP, __BLK_TA_REQUEUE, __BLK_TA_SLEEPRQ, __BLK_TA_SPLIT,
    __BLK_TA_UNPLUG_IO, __BLK_TA_UNPLUG_TIMER,
};
use crate::bindings::{BLK_IO_TRACE_MAGIC, BLK_TC_DISCARD, BLK_TC_READ, BLK_TC_SYNC, BLK_TC_WRITE};
use crate::BlkIoTrace;

// Manually define constants and macros that bindgen doesn't handle well.
pub const BLK_TC_SHIFT: u32 = 16;
#[allow(non_snake_case)]
pub const fn BLK_TC_ACT(act: u32) -> u32 {
    act << BLK_TC_SHIFT
}
pub const __BLK_TA_QUEUE: u32 = 1;

#[derive(Clone, Copy, PartialEq)]
pub enum Endianness {
    Unknown,
    Native,
    Big,
    Little,
}

pub fn check_data_endianness(magic: u32) -> Endianness {
    if (magic & 0xffffff00) as u32 == BLK_IO_TRACE_MAGIC {
        Endianness::Native
    } else if u32::from_be(magic) & 0xffffff00 == BLK_IO_TRACE_MAGIC {
        Endianness::Big
    } else if u32::from_le(magic) & 0xffffff00 == BLK_IO_TRACE_MAGIC {
        Endianness::Little
    } else {
        Endianness::Unknown
    }
}

pub fn correct_endianness(trace: &mut BlkIoTrace, endianness: Endianness) {
    match endianness {
        Endianness::Native => return,
        Endianness::Little => {
            trace.magic = u32::from_le(trace.magic);
            trace.sequence = u32::from_le(trace.sequence);
            trace.time = u64::from_le(trace.time);
            trace.sector = u64::from_le(trace.sector);
            trace.bytes = u32::from_le(trace.bytes);
            trace.action = u32::from_le(trace.action);
            trace.pid = u32::from_le(trace.pid);
            trace.device = u32::from_le(trace.device);
            trace.cpu = u32::from_le(trace.cpu);
            trace.error = u16::from_le(trace.error);
            trace.pdu_len = u16::from_le(trace.pdu_len);
        }
        Endianness::Big => {
            trace.magic = u32::from_be(trace.magic);
            trace.sequence = u32::from_be(trace.sequence);
            trace.time = u64::from_be(trace.time);
            trace.sector = u64::from_be(trace.sector);
            trace.bytes = u32::from_be(trace.bytes);
            trace.action = u32::from_be(trace.action);
            trace.pid = u32::from_be(trace.pid);
            trace.device = u32::from_be(trace.device);
            trace.cpu = u32::from_be(trace.cpu);
            trace.error = u16::from_be(trace.error);
            trace.pdu_len = u16::from_be(trace.pdu_len);
        }
        Endianness::Unknown => panic!(),
    }
}

pub fn format_trace(trace: &BlkIoTrace) -> String {
    let rw = if (trace.action & BLK_TC_ACT(BLK_TC_WRITE)) != 0 {
        "W"
    } else if (trace.action & BLK_TC_ACT(BLK_TC_READ)) != 0 {
        "R"
    } else if (trace.action & BLK_TC_ACT(BLK_TC_DISCARD)) != 0 {
        "D"
    } else if (trace.action & BLK_TC_ACT(BLK_TC_SYNC)) != 0 {
        "S"
    } else {
        "N"
    };

    #[allow(non_snake_case)]
    let action_char = match trace.action & 0x0000ffff {
        __BLK_TA_QUEUE => 'Q',
        __BLK_TA_BACKMERGE => 'M',
        __BLK_TA_FRONTMERGE => 'F',
        __BLK_TA_GETRQ => 'G',
        __BLK_TA_SLEEPRQ => 'S',
        __BLK_TA_REQUEUE => 'R',
        __BLK_TA_ISSUE => 'I',
        __BLK_TA_COMPLETE => 'C',
        __BLK_TA_PLUG => 'P',
        __BLK_TA_UNPLUG_IO => 'U',
        __BLK_TA_UNPLUG_TIMER => 'T',
        __BLK_TA_INSERT => 'A',
        __BLK_TA_SPLIT => 'S',
        __BLK_TA_BOUNCE => 'B',
        __BLK_TA_REMAP => 'm',
        __BLK_TA_ABORT => 'X',
        __BLK_TA_DRV_DATA => 'D',
        _ => '?',
    };

    format!(
        "{},{} {} {}.{:09} {} {} {} {} + {}",
        trace.cpu,
        trace.pid,
        trace.sequence,
        trace.time / 1_000_000_000,
        trace.time % 1_000_000_000,
        action_char,
        rw,
        trace.error,
        trace.sector,
        trace.bytes
    )
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::bindings::{BLK_IO_TRACE_MAGIC, BLK_IO_TRACE_VERSION, BLK_TC_WRITE};
    use crate::BlkIoTrace;

    fn get_test_trace(sequence: u32, time: u64) -> BlkIoTrace {
        BlkIoTrace {
            magic: BLK_IO_TRACE_MAGIC | BLK_IO_TRACE_VERSION,
            sequence,
            time,
            sector: 1024,
            bytes: 4096,
            action: (__BLK_TA_QUEUE | BLK_TC_ACT(BLK_TC_WRITE)),
            pid: 123,
            device: (253 << 20 | 1),
            cpu: 2,
            error: 0,
            pdu_len: 0,
        }
    }

    #[test]
    fn test_format() {
        let trace = get_test_trace(1, 1234567890);
        let formatted = format_trace(&trace);
        assert_eq!(formatted, "2,123 1 1.234567890 Q W 0 1024 + 4096");
    }

    #[test]
    fn test_endianness_correction() {
        let mut trace = get_test_trace(1, 200);
        let original_trace = trace.clone();

        let test_endianness = if cfg!(target_endian = "big") {
            Endianness::Little
        } else {
            Endianness::Big
        };

        correct_endianness(&mut trace, test_endianness);
        assert_ne!(original_trace.magic, trace.magic, "magic number should change after endianness correction");
        correct_endianness(&mut trace, test_endianness);
        assert_eq!(original_trace.magic, trace.magic, "magic number should be restored after second endianness correction");
    }
}
