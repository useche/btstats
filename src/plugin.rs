use crate::blk_io_trace::BlkIoTrace;
use std::string::ToString;

trait Plugin {
    /// Name of the plugin.
    fn name() -> impl ToString;

    /// Updates the plugin's aggregations with this trace.
    fn update(trace: &BlkIoTrace);

    /// Returns a string with the accumulated statistics so far. They will be formatted with
    /// Markdown, so they can be printed directly to stdout.
    fn result() -> impl ToString;
}
