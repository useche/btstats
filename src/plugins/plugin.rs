use crate::blk_io_trace::BlkIoTrace;
use std::string::ToString;

pub trait Plugin {
    /// Name of the plugin.
    fn name() -> impl ToString;

    /// Updates the plugin's aggregations with this trace.
    fn update(&mut self, trace: &BlkIoTrace);

    /// Returns a string with the accumulated statistics so far. They will be formatted with
    /// Markdown, so they can be printed directly to stdout.
    fn result(&self) -> impl ToString;
}
