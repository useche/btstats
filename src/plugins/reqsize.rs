use super::plugin::Plugin;
use crate::blk_io_trace::{BlkIoTrace, IoAction, IoCategory};
use std::cmp::{max, min};

pub struct ReqSize {
    min: u32,
    max: u32,
    total_size_blks: u64,
    num_reads: u64,
    num_writes: u64,
    num_discards: u64,
}

impl Default for ReqSize {
    fn default() -> Self {
        Self {
            min: u32::MAX,
            max: 0,
            total_size_blks: 0,
            num_reads: 0,
            num_writes: 0,
            num_discards: 0,
        }
    }
}

impl Plugin for ReqSize {
    fn name(&self) -> String {
        "Request Size".to_string()
    }

    fn update(&mut self, trace: &BlkIoTrace) {
        if trace.io_action() != IoAction::Complete {
            return;
        }

        let blks = trace.blks();

        if blks == 0 {
            return;
        }

        match trace.io_category() {
            IoCategory::Read => self.num_reads += 1,
            IoCategory::Write => self.num_writes += 1,
            IoCategory::Discard => self.num_discards += 1,
            _ => return,
        };

        self.min = min(self.min, blks);
        self.max = max(self.max, blks);
        self.total_size_blks += blks as u64;
    }

    fn result(&self) -> String {
        let total_reqs = self.num_reads + self.num_writes;

        if total_reqs == 0 {
            return "No updates".to_string();
        }

        format!(
            "Reqs. #: {} Reads: {} ({:.2}%) Writes: {} ({:.2}%) Size:(min: {} avg: {:.2} max: {} (blks))",
            total_reqs,
            self.num_reads,
            self.num_reads as f64 / total_reqs as f64,
            self.num_writes,
            self.num_writes as f64 / total_reqs as f64,
            self.min,
            self.total_size_blks as f64 / total_reqs as f64,
            self.max,
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::bindings::{
        blk_io_trace, BLK_IO_TRACE_MAGIC, BLK_IO_TRACE_VERSION, BLK_TC_DISCARD, BLK_TC_READ,
        BLK_TC_WRITE, __BLK_TA_COMPLETE, __BLK_TA_QUEUE,
    };

    const BLK_TC_SHIFT: u32 = 16;
    const fn blk_tc_act(act: u32) -> u32 {
        act << BLK_TC_SHIFT
    }

    fn create_trace(action: u32, bytes: u32) -> BlkIoTrace {
        let trace = blk_io_trace {
            magic: BLK_IO_TRACE_MAGIC | BLK_IO_TRACE_VERSION,
            sequence: 0,
            time: 0,
            sector: 0,
            bytes,
            action,
            pid: 0,
            device: 0,
            cpu: 0,
            error: 0,
            pdu_len: 0,
        };
        BlkIoTrace::from_bytes(unsafe { std::mem::transmute(trace) }).unwrap()
    }

    #[test]
    fn update_with_complete_traces() {
        let mut req_size = ReqSize::default();

        let read_trace = create_trace(__BLK_TA_COMPLETE | blk_tc_act(BLK_TC_READ), 512);
        let write_trace = create_trace(__BLK_TA_COMPLETE | blk_tc_act(BLK_TC_WRITE), 1024);
        let discard_trace = create_trace(__BLK_TA_COMPLETE | blk_tc_act(BLK_TC_DISCARD), 2048);

        req_size.update(&read_trace);
        req_size.update(&write_trace);
        req_size.update(&discard_trace);

        assert_eq!(req_size.num_reads, 1);
        assert_eq!(req_size.num_writes, 1);
        assert_eq!(req_size.num_discards, 1);
        assert_eq!(req_size.min, 1);
        assert_eq!(req_size.max, 4);
        assert_eq!(req_size.total_size_blks, 7);
    }

    #[test]
    fn update_with_non_complete_traces() {
        let mut req_size = ReqSize::default();
        let trace = create_trace(__BLK_TA_QUEUE | blk_tc_act(BLK_TC_READ), 512);
        req_size.update(&trace);

        assert_eq!(req_size.num_reads, 0);
        assert_eq!(req_size.num_writes, 0);
        assert_eq!(req_size.num_discards, 0);
        assert_eq!(req_size.total_size_blks, 0);
    }

    #[test]
    fn update_with_zero_blocks() {
        let mut req_size = ReqSize::default();
        let trace = create_trace(__BLK_TA_COMPLETE | blk_tc_act(BLK_TC_READ), 0);
        req_size.update(&trace);

        assert_eq!(req_size.num_reads, 0);
        assert_eq!(req_size.num_writes, 0);
        assert_eq!(req_size.num_discards, 0);
        assert_eq!(req_size.total_size_blks, 0);
    }

    #[test]
    fn result_output() {
        let mut req_size = ReqSize::default();

        let read_trace = create_trace(__BLK_TA_COMPLETE | blk_tc_act(BLK_TC_READ), 512);
        let write_trace = create_trace(__BLK_TA_COMPLETE | blk_tc_act(BLK_TC_WRITE), 1024);

        req_size.update(&read_trace);
        req_size.update(&write_trace);

        let result = req_size.result().to_string();
        assert_eq!(
            result,
            "Reqs. #: 2 Reads: 1 (0.50%) Writes: 1 (0.50%) Size:(min: 1 avg: 1.50 max: 2 (blks))"
        );
    }

    #[test]
    fn test_no_updates() {
        let req_size = ReqSize::default();

        assert_eq!(
            req_size.result(),
            "No updates"
        );
    }
}
