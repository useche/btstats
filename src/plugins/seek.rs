use super::plugin::Plugin;
use crate::blk_io_trace::{BlkIoTrace, IoAction};
use std::cmp::{max, min};

pub struct Seek {
    last_pos: Option<u64>,
    min: u64,
    max: u64,
    total_distance: u64,
    seeks: u64,
    total_reqs: u64,
}

impl Default for Seek {
    fn default() -> Self {
        Self {
            last_pos: None,
            min: u64::MAX,
            max: 0,
            total_distance: 0,
            seeks: 0,
            total_reqs: 0,
        }
    }
}

impl Plugin for Seek {
    fn name(&self) -> String {
        "Seek".to_string()
    }

    fn update(&mut self, trace: &BlkIoTrace) {
        if trace.io_action() != IoAction::Complete {
            return;
        }

        self.total_reqs += 1;
        let blks = trace.blks() as u64;
        let sector = trace.trace().sector;

        if let Some(last_pos) = self.last_pos {
            if last_pos != sector {
                let distance = if last_pos > sector {
                    last_pos - sector
                } else {
                    sector - last_pos
                };
                self.total_distance += distance;
                self.max = max(self.max, distance);
                self.min = min(self.min, distance);
                self.seeks += 1;
            }
        }

        self.last_pos = Some(sector + blks);
    }

    fn result(&self) -> String {
        if self.seeks == 0 {
            return "No updates".to_string();
        }

        format!(
            "Seq.: {:.2}% Seeks #: {} min: {} avg: {:.2} max: {} (blks)",
            (1.0 - (self.seeks as f64 / self.total_reqs as f64)) * 100.0,
            self.seeks,
            self.min,
            self.total_distance as f64 / self.seeks as f64,
            self.max
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::bindings::{
        blk_io_trace, BLK_IO_TRACE_MAGIC, BLK_IO_TRACE_VERSION, BLK_TC_READ,
        __BLK_TA_COMPLETE,
    };

    const BLK_TC_SHIFT: u32 = 16;
    const fn blk_tc_act(act: u32) -> u32 {
        act << BLK_TC_SHIFT
    }

    fn create_trace(action: u32, bytes: u32, sector: u64) -> BlkIoTrace {
        let trace = blk_io_trace {
            magic: BLK_IO_TRACE_MAGIC | BLK_IO_TRACE_VERSION,
            sequence: 0,
            time: 0,
            sector,
            bytes,
            action,
            pid: 0,
            device: 0,
            cpu: 0,
            error: 0,
            pdu_len: 0,
        };
        BlkIoTrace::from_bytes(unsafe { std::mem::transmute(trace) }).unwrap()
    }

    #[test]
    fn test_seeks() {
        let mut seek = Seek::default();

        // First request, no seek yet
        let trace1 = create_trace(__BLK_TA_COMPLETE | blk_tc_act(BLK_TC_READ), 512, 0);
        seek.update(&trace1);
        assert_eq!(seek.seeks, 0);
        assert_eq!(seek.total_reqs, 1);
        assert_eq!(seek.last_pos, Some(1));

        // Second request, with a seek
        let trace2 = create_trace(__BLK_TA_COMPLETE | blk_tc_act(BLK_TC_READ), 512, 10);
        seek.update(&trace2);
        assert_eq!(seek.seeks, 1);
        assert_eq!(seek.total_reqs, 2);
        assert_eq!(seek.total_distance, 9);
        assert_eq!(seek.min, 9);
        assert_eq!(seek.max, 9);
        assert_eq!(seek.last_pos, Some(11));

        // Third request, sequential (no seek)
        let trace3 = create_trace(__BLK_TA_COMPLETE | blk_tc_act(BLK_TC_READ), 512, 11);
        seek.update(&trace3);
        assert_eq!(seek.seeks, 1);
        assert_eq!(seek.total_reqs, 3);
        assert_eq!(seek.last_pos, Some(12));

        // Fourth request, with a seek
        let trace4 = create_trace(__BLK_TA_COMPLETE | blk_tc_act(BLK_TC_READ), 512, 2);
        seek.update(&trace4);
        assert_eq!(seek.seeks, 2);
        assert_eq!(seek.total_reqs, 4);
        assert_eq!(seek.total_distance, 19);
        assert_eq!(seek.min, 9);
        assert_eq!(seek.max, 10);
        assert_eq!(seek.last_pos, Some(3));
    }

    #[test]
    fn test_no_seeks() {
        let mut seek = Seek::default();

        let trace1 = create_trace(__BLK_TA_COMPLETE | blk_tc_act(BLK_TC_READ), 512, 0);
        seek.update(&trace1);

        let trace2 = create_trace(__BLK_TA_COMPLETE | blk_tc_act(BLK_TC_READ), 1024, 1);
        seek.update(&trace2);

        assert_eq!(seek.seeks, 0);
        assert_eq!(seek.total_reqs, 2);
        assert_eq!(seek.total_distance, 0);
        assert_eq!(seek.min, u64::MAX);
        assert_eq!(seek.max, 0);
    }

    #[test]
    fn test_no_updates() {
        let seek = Seek::default();
        assert_eq!(seek.result(), "No updates");
    }

    #[test]
    fn test_result_output() {
        let mut seek = Seek::default();

        let trace1 = create_trace(__BLK_TA_COMPLETE | blk_tc_act(BLK_TC_READ), 512, 0);
        seek.update(&trace1);
        let trace2 = create_trace(__BLK_TA_COMPLETE | blk_tc_act(BLK_TC_READ), 512, 10);
        seek.update(&trace2);

        let result = seek.result();
        assert_eq!(
            result,
            "Seq.: 50.00% Seeks #: 1 min: 9 avg: 9.00 max: 9 (blks)"
        );
    }
}