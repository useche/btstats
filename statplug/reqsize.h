#include <asm/types.h>
#include <plugins.h>

#define DECL_ASSIGN_REQSIZE(name,data)		\
	struct reqsize_data *name = (struct reqsize_data *)data

struct reqsize_data
{
	__u64 min;
	__u64 max;
	__u64 total_size;
	__u64 reqs;
	__u64 reads;
};
