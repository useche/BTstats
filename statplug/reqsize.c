/*
 * Author: Luis Useche (August 2008)
 * Email: luis@cs.fiu.edu
 *
 * BSD License
 * Copyright (c) 2008, Luis Useche
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *
 *     * Redistributions in binary form must reproduce the above
 *       copyright notice, this list of conditions and the following
 *       disclaimer in the documentation and/or other materials
 *       provided with the distribution.
 *
 *     * Neither the name of Luis Useche nor the names of its
 *       contributors may be used to endorse or promote products
 *       derived from this software without specific prior written
 *       permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
 * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
 * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include <asm/types.h>
#include <glib.h>
#include <stdio.h>

#include <blktrace_api.h>
#include <blktrace.h>
#include <plugins.h>
#include <utils.h>

#include <reqsize.h>

static void C(struct blk_io_trace *t, void *data)
{
	DECL_ASSIGN_REQSIZE(rsd,data);
	
	__u64 blks = t_blks(t);
	
	if(blks) {
		rsd->min = MIN(rsd->min, blks);
		rsd->max = MAX(rsd->max, blks);
		rsd->total_size += blks;
		rsd->reqs++;
	}
}

void reqsize_add(void *data1, const void *data2) 
{
	DECL_ASSIGN_REQSIZE(rsd1,data1);
	DECL_ASSIGN_REQSIZE(rsd2,data2);
	
	rsd1->min = MIN(rsd1->min,rsd2->min);
	rsd1->max = MAX(rsd1->max,rsd2->max);
	rsd1->total_size += rsd2->total_size;
	rsd1->reqs += rsd2->reqs;
}

void reqsize_print_results(const void *data)
{
	DECL_ASSIGN_REQSIZE(rsd,data);

	if(rsd->reqs)
		printf("Reqs. #: %lld min: %lld avg: %f max: %lld (blocks)\n",
		       rsd->reqs,
		       rsd->min,
		       ((double)rsd->total_size)/rsd->reqs,
		       rsd->max);
}

void reqsize_init(struct plugin *p, struct plugin_set *__unused)
{
	struct reqsize_data *req = p->data = g_new(struct reqsize_data,1);
	req->min = ~0;
	req->max = 0;
	req->total_size = 0;
	req->reqs = 0;
	
	__unused = NULL; /* to make gcc quite */
}

void reqsize_ops_init(struct plugin_ops *po)
{
	po->add = reqsize_add;
	po->print_results = reqsize_print_results;
	
	/* association of event int and function */
	g_tree_insert(po->event_tree,(gpointer)__BLK_TA_COMPLETE,C);
}

void reqsize_destroy(struct plugin *p)
{
	g_free(p->data);
}
