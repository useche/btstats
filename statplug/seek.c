#include <asm/types.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

#include <blktrace.h>
#include <blktrace_api.h>
#include <list_plugins.h>
#include <plugins.h>
#include <utils.h>

#include <reqsize.h>

#define DECL_ASSIGN_SEEK(name, data)                                           \
  struct seek_data *name = (struct seek_data *)data

struct seek_data {
  __u64 lastpos;
  struct reqsize_data *req_dat;

  __u64 max;
  __u64 min;
  __u64 total;
  __u64 seeks;
};

static void C(const struct blk_io_trace *t, void *data) {
  DECL_ASSIGN_SEEK(seek, data);

  __u64 blks = t_blks(t);

  if (seek->lastpos != UINT64_MAX) {
    if (seek->lastpos != t->sector) {
      __u64 distance = seek->lastpos > t->sector ? seek->lastpos - t->sector
                                                 : t->sector - seek->lastpos;
      seek->total += distance;
      seek->max = MAX(seek->max, distance);
      seek->min = MIN(seek->min, distance);
      seek->seeks++;
    }
  }

  seek->lastpos = t->sector + blks;
}

void seek_add(void *data1, const void *data2) {
  DECL_ASSIGN_SEEK(seek1, data1);
  DECL_ASSIGN_SEEK(seek2, data2);

  seek1->min = MIN(seek1->min, seek2->min);
  seek1->max = MAX(seek2->max, seek2->max);
  seek1->total += seek2->total;
  seek1->seeks += seek2->seeks;
}

void seek_print_results(const void *data) {
  DECL_ASSIGN_SEEK(seek, data);

  if (seek->seeks) {
    printf("Seq.: %.2f%%\n",
           (1 - (((double)seek->seeks) / seek->req_dat->total_size)) * 100);
    printf(
        "Seeks #: %llu min: %llu avg: %f max: %llu (blks)\n",
        (long long unsigned int)seek->seeks, (long long unsigned int)seek->min,
        ((double)seek->total) / seek->seeks, (long long unsigned int)seek->max);
  }
}

void seek_init(struct plugin *p, struct plugin_set *ps,
               struct plug_args *__un) {
  struct seek_data *seek = p->data = calloc(1, sizeof(struct seek_data));
  seek->lastpos = UINT64_MAX;
  seek->max = 0;
  seek->min = ~0;
  seek->total = 0;
  seek->seeks = 0;
  seek->req_dat = (struct reqsize_data *)ps->plugs[REQ_SIZE_IND].data;
}

void seek_destroy(struct plugin *p) { free(p->data); }

void seek_ops_init(struct plugin_ops *po) {
  po->add = seek_add;
  po->print_results = seek_print_results;

  struct event_entry *e = malloc(sizeof(struct event_entry));
  e->event_key = __BLK_TA_COMPLETE;
  e->event_handler = (event_func_t)C;
  RB_INSERT(event_tree_head, po->event_tree, e);
}
