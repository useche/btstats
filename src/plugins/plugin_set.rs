use crate::blk_io_trace::BlkIoTrace;

use super::plugin::Plugin;
use super::reqsize::ReqSize;
use super::seek::Seek;

pub struct PluginSet {
    plugins: Vec<Box<dyn Plugin>>,
}

impl Default for PluginSet {
    fn default() -> Self {
        Self {
            plugins: vec![Box::new(ReqSize::default()), Box::new(Seek::default())],
        }
    }
}

impl PluginSet {
    pub fn update(&mut self, trace: &BlkIoTrace) {
        self.plugins.iter_mut().for_each(|p| p.update(trace));
    }

    pub fn result(&self) -> String {
        self.plugins.iter().fold(String::new(), |acc, p| {
            let p_header = format!("## {}", p.name());
            let p_result = p.result().to_string();

            format!("{}{}\n{}\n", acc, p_header, p_result)
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::bindings::{
        blk_io_trace, BLK_IO_TRACE_MAGIC, BLK_IO_TRACE_VERSION, BLK_TC_DISCARD, BLK_TC_READ,
        BLK_TC_WRITE, __BLK_TA_COMPLETE, __BLK_TA_QUEUE,
    };

    const BLK_TC_SHIFT: u32 = 16;
    const fn blk_tc_act(act: u32) -> u32 {
        act << BLK_TC_SHIFT
    }

    fn create_trace(action: u32, bytes: u32, sector: u64) -> BlkIoTrace {
        let trace = blk_io_trace {
            magic: BLK_IO_TRACE_MAGIC | BLK_IO_TRACE_VERSION,
            sequence: 0,
            time: 0,
            sector,
            bytes,
            action,
            pid: 0,
            device: 0,
            cpu: 0,
            error: 0,
            pdu_len: 0,
        };
        BlkIoTrace::from_bytes(unsafe { std::mem::transmute(trace) }).unwrap()
    }

    #[test]
    fn test_plugin_set_no_updates() {
        let plugin_set = PluginSet::default();
        insta::assert_snapshot!(plugin_set.result())
    }

    #[test]
    fn test_plugin_set_update_with_non_complete() {
        let mut plugin_set = PluginSet::default();
        let trace = create_trace(__BLK_TA_QUEUE | blk_tc_act(BLK_TC_READ), 512, 0);
        plugin_set.update(&trace);
        insta::assert_snapshot!(plugin_set.result())
    }

    #[test]
    fn test_plugin_set_one_update_with_complete() {
        let mut plugin_set = PluginSet::default();
        let trace = create_trace(__BLK_TA_COMPLETE | blk_tc_act(BLK_TC_READ), 512, 0);
        plugin_set.update(&trace);
        insta::assert_snapshot!(plugin_set.result())
    }

    #[test]
    fn test_plugin_set_update_with_complete() {
        let mut plugin_set = PluginSet::default();
        let traces = vec![
            create_trace(__BLK_TA_COMPLETE | blk_tc_act(BLK_TC_READ), 512, 0),
            create_trace(__BLK_TA_COMPLETE | blk_tc_act(BLK_TC_WRITE), 1024, 10),
            create_trace(__BLK_TA_COMPLETE | blk_tc_act(BLK_TC_DISCARD), 2048, 2),
            create_trace(__BLK_TA_COMPLETE | blk_tc_act(BLK_TC_READ), 2048, 12),
        ];

        for trace in &traces {
            plugin_set.update(trace);
        }

        insta::assert_snapshot!(plugin_set.result())
    }
}
