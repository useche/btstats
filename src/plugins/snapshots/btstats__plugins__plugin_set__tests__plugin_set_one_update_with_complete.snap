---
source: src/plugins/plugin_set.rs
expression: plugin_set.result()
---
## Request Size
Reqs. #: 1 Reads: 1 (100.00%) Writes: 0 (0.00%) Discard: 0 (0.00%) Size:(min: 1 avg: 1.00 max: 1 (blks))
## Seek
No updates
